### PR TITLE
Add start and stop options for blackduck in scope of issue #354

### LINKED ISSUE #354: Deprecate desiredstate option, and add start/stop commands
**What would you like to be added**:
desiredstate is an implementation detail in the CRD.  Should remove that option from synopsycctl and replace with --start and --stop options.

**Why is this needed**:
Method by which blackducks are started and stopped may change.

### PR DESCRIPTION
**Pull request by yashbhutwala**
Closes #354 

/assign @rrati

### CODE REVIEW

**Comment by operator-r2d2** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/blackducksoftware/synopsys-operator/pull/357#" title="Author self-approved">yashbhutwala</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=blackducksoftware%2Fsynopsys-operator).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/blackducksoftware/synopsys-operator/blob/master/OWNERS)~~ [yashbhutwala]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: pkg/blackduck/ctl_blackduck.go [hunk 1]
```diff
@@ -340,8 +341,14 @@ func (ctl *Ctl) SetFlag(f *pflag.Flag) {
 			ctl.Spec.AuthCustomCA = data
 		case "type":
 			ctl.Spec.Type = ctl.Type
-		case "desired-state":
-			ctl.Spec.DesiredState = ctl.DesiredState
+		case "start":
```
Comment (msenmurugan) [inline pkg/blackduck/ctl_blackduck.go]: what will happen if both are set? :-)

File: pkg/blackduck/ctl_blackduck.go [hunk 2]
```diff
@@ -218,7 +218,8 @@ func (ctl *Ctl) AddSpecFlags(cmd *cobra.Command, master bool) {
 	cmd.Flags().StringVar(&ctl.ProxyCertificateFilePath, "proxy-certificate-file-path", ctl.ProxyCertificateFilePath, "Absolute path to a file for the Black Duck proxy certificate")
 	cmd.Flags().StringVar(&ctl.AuthCustomCAFilePath, "auth-custom-ca-file-path", ctl.AuthCustomCAFilePath, "Absolute path to a file for the Custom Auth CA for Black Duck")
 	cmd.Flags().StringVar(&ctl.Type, "type", ctl.Type, "Type of Blackduck")
-	cmd.Flags().StringVar(&ctl.DesiredState, "desired-state", ctl.DesiredState, "Desired state of Blackduck")
+	cmd.Flags().BoolVar(&ctl.Start, "start", ctl.Start, "Start the Blackduck")
```
Comment (msenmurugan) [inline pkg/blackduck/ctl_blackduck.go]: @rrati whether there should be any default behavior?

File: pkg/blackduck/ctl_blackduck.go (hunk 1 above)
Comment (yashbhutwala) [reply pkg/blackduck/ctl_blackduck.go]: Maybe there should be only one flag?  What does it mean to have `stop` = false?

File: pkg/blackduck/ctl_blackduck.go [hunk 3]
```diff
@@ -65,7 +65,8 @@ type Ctl struct {
 	ProxyCertificateFilePath              string
 	AuthCustomCAFilePath                  string
 	Type                                  string
-	DesiredState                          string
+	Start                                 bool
```
Comment (rrati) [inline pkg/blackduck/ctl_blackduck.go]: Leave DesiredState in the CRD/ctl structs.  Only thing that needs to change is the outward appearance.  --stop and --start just set the desiredState field appropriately.

File: pkg/blackduck/ctl_blackduck.go [hunk 4]
```diff
@@ -136,6 +136,13 @@ func isValidSize(size string) bool {
 
 // CheckSpecFlags returns an error if a user input was invalid
 func (ctl *Ctl) CheckSpecFlags() error {
+	// log.Printf("ctl.Start: %v \n", ctl.Start)
+	// log.Printf("ctl.Stop: %v \n", ctl.Stop)
+	// log.Printf("ctl.Spec.DesiredState: %v \n", ctl.Spec.DesiredState)
+	if true == ctl.Start && true == ctl.Stop {
+		return fmt.Errorf("Both start and stop cannot be same value")
```
Comment (rrati) [inline pkg/blackduck/ctl_blackduck.go]: This message is a bit confusing.  Suggesting something like "invalid options.  Only one of start or stop is allowed"

File: pkg/blackduck/ctl_blackduck.go (hunk 2 above)
Comment (rrati) [reply pkg/blackduck/ctl_blackduck.go]: No.  If no options are given we don't want to change anything, nor is there a good default value since we don't want to do anything if neither option is set.

File: pkg/blackduck/ctl_blackduck.go [hunk 5]
```diff
@@ -340,8 +348,13 @@ func (ctl *Ctl) SetFlag(f *pflag.Flag) {
 			ctl.Spec.AuthCustomCA = data
 		case "type":
 			ctl.Spec.Type = ctl.Type
-		case "desired-state":
-			ctl.Spec.DesiredState = ctl.DesiredState
+		case "start", "stop":
```
Comment (rrati) [inline pkg/blackduck/ctl_blackduck.go]: No need for the complexity here.  We've disallowed both start and stop to be provided together, so just set the spec.DesiredState to START or STOP in separate case statements.  ie

case "start":
  ctl.Spec.DesiredState = "START"
case "stop":
  ctl.Spec.DesiredState = "STOP"

File: pkg/blackduck/ctl_blackduck_test.go [hunk 6]
```diff
@@ -389,13 +389,25 @@ func TestSetFlag(t *testing.T) {
 		},
 		// case
 		{
-			flagName:   "desired-state",
+			flagName:   "start",
```
Comment (rrati) [inline pkg/blackduck/ctl_blackduck_test.go]: Add a test to verify start and stop is disallowed on on the same command.

File: pkg/blackduck/ctl_blackduck.go [hunk 7]
```diff
@@ -136,6 +136,13 @@ func isValidSize(size string) bool {
 
 // CheckSpecFlags returns an error if a user input was invalid
 func (ctl *Ctl) CheckSpecFlags() error {
+	// log.Printf("ctl.Start: %v \n", ctl.Start)
+	// log.Printf("ctl.Stop: %v \n", ctl.Stop)
+	// log.Printf("ctl.Spec.DesiredState: %v \n", ctl.Spec.DesiredState)
```
Comment (rrati) [inline pkg/blackduck/ctl_blackduck.go]: Remove commented out code

File: pkg/synopsysctl/cmd_start.go [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+var startCmd = &cobra.Command{
+	Use:   "start [resource]",
+	Short: "Start a Synopsys Resource in your cluster",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		return fmt.Errorf("Must specify a resource")
+	},
+}
+
+// startBlackduckCmd starts a Blackduck in the cluster
+var startBlackduckCmd = &cobra.Command{
```
Comment (msenmurugan) [inline pkg/synopsysctl/cmd_start.go]: @yashbhutwala 
Will the start and stop for opssight and alert will be in separate PR?

Comment (yashbhutwala) [reply pkg/synopsysctl/cmd_start.go]: Done.

File: pkg/apps/alert/latest/alert_creater.go [hunk 9]
```diff
@@ -60,11 +62,20 @@ func (ac *Creater) Ensure(alert *alertapi.Alert) error {
...
 		return err
 	}
-	// Update components in cluster
-	commonConfig := crdupdater.NewCRUDComponents(ac.KubeConfig, ac.KubeClient, ac.Config.DryRun, false, alert.Spec.Namespace, cpList, "app=alert")
-	_, errors := commonConfig.CRUDComponents()
-	if len(errors) > 0 {
-		return fmt.Errorf("unable to update Alert components due to %+v", errors)
+	if strings.EqualFold(alert.Spec.DesiredState, "STOP") {
+		commonConfig := crdupdater.NewCRUDComponents(ac.KubeConfig, ac.KubeClient, ac.Config.DryRun, false, alert.Spec.Namespace,
+			&api.ComponentList{PersistentVolumeClaims: cpList.PersistentVolumeClaims}, "app=alert")
```
Comment (yashbhutwala) [inline pkg/apps/alert/latest/alert_creater.go]: This is mimicking blackduck stop for now.  @jdartigalongue @msenmurugan and I discussed that this should be made same as opssight for both blackduck and alert, in that when desired state is stop, scaled down everything to 0 as well as don't delete the PVC.  What do you think?

File: cmd/operator-ui/templates/blackducks/_form.html [hunk 10]
```diff
@@ -1,81 +1,81 @@
...
-  <div id="pvcId">
-    <%= f.SelectTag("Spec.PVCStorageClass", {value: blackduck.Spec.PVCStorageClass, label: "PVC Storage Class", options: blackduck.View.StorageClasses}) %>
-    <%= if (blackduck.Spec.Namespace!="" && blackduck.Spec.PersistentStorage==true) { %>
+  <div class="new-container">
+    <%= f.InputTag("Spec.Namespace", {value: blackduck.Spec.Namespace, label: "Namespace *", required: true, placeholder: "Lowercase letters, numbers, and hyphens only. Start with a lowercase letter and do not end with a hyphen", pattern: "^[a-z][-a-z0-9]+", title: "Lowercase letter, numbers, and hyphen only. Start with lowercase letter and cannot start or end with hyphen."}) %>
+    <%= f.SelectTag("Spec.Version", {options: blackduck.View.SupportedVersions, value: blackduck.Spec.Version, label: "Black Duck Version"}) %>
+    <%= f.InputTag("Spec.LicenseKey", {value: blackduck.Spec.LicenseKey, label: "License Key"}) %>
+    <%= if (blackduck.Spec.Namespace!="") { %>
+    <%= f.SelectTag("Spec.DesiredState", {options: ["", "Stop"], value: blackduck.Spec.DesiredState, label: "Desired State"}) %>
```
Comment (msenmurugan) [inline cmd/operator-ui/templates/blackducks/_form.html]: @yashbhutwala whether it will not confuse the user that only blank and stop are displayed?

File: examples/deprecated/opssight.json [hunk 11]
```diff
@@ -6,7 +6,7 @@
   },
   "spec": {
     "namespace": "opssight-test",
-    "desiredState": "start",
+    "desiredState": "",
```
Comment (msenmurugan) [inline examples/deprecated/opssight.json]: I think opssight won't work with empty desiredState.. I think @yashbhutwala need to revert the change

File: pkg/apps/alert/latest/alert_creater.go (hunk 9 above)
Comment (jdartigalongue) [reply pkg/apps/alert/latest/alert_creater.go]: I personally prefer deleting the resources when the instance is stopped but I don't have strong feelings about this. What are the advantages of scaling down everything to 0 over the way it currently works?

Comment (yashbhutwala) [reply pkg/apps/alert/latest/alert_creater.go]: Keeping as is for now.

File: cmd/operator-ui/templates/blackducks/_form.html (hunk 10 above)
Comment (yashbhutwala) [reply cmd/operator-ui/templates/blackducks/_form.html]: Done.

File: pkg/opssight/handler.go [hunk 12]
```diff
@@ -61,7 +61,7 @@ const (
 	Error State = "Error"
 
 	// Start is used when OpsSight deployment to be created or updated
-	Start DesiredState = "Start"
+	Start DesiredState = ""
```
[the remark this answers is not in the document]
Comment (yashbhutwala) [reply pkg/opssight/handler.go]: Called only once, I just tested it, and start and stop works for opssight.

**Comment by operator-r2d2** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/blackducksoftware/synopsys-operator/pull/357#pullrequestreview-243870601" title="Approved">msenmurugan</a>*, *<a href="https://github.com/blackducksoftware/synopsys-operator/pull/357#" title="Author self-approved">yashbhutwala</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=blackducksoftware%2Fsynopsys-operator).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/blackducksoftware/synopsys-operator/blob/master/OWNERS)~~ [msenmurugan,yashbhutwala]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
